daemon/api_icons.go

```diff
@@ -69,3 +86,106 @@ func iconGet(st *state.State, name string) Response {
 
 	return fileResponse(icon)
 }
+
+var (
+	desktopSection       = "Desktop Entry"
+	localizedNameMatcher = regexp.MustCompile(`^Name\[(\w+)\]$`)
+)
+
+func snapAppIconGet(c *Command, r *http.Request, user *auth.UserState) Response {
+	vars := muxVars(r)
+	snap := vars["snap"]
+	app := vars["app"]
+
+	snapInfo, err := localSnapInfo(c.d.overlord.State(), snap)
+	if err != nil {
+		if errors.Is(err, errNoSnap) {
+			return SnapNotFound(snap, err)
+		}
+		return InternalError(fmt.Sprintf("%v", err))
+	}
+
+	for _, appInfo := range snapInfo.info.Apps {
+		if appInfo.Name != app {
+			continue
+		}
+
+		parser := goconfigparser.New()
+		if err := parser.ReadFile(appInfo.DesktopFile()); err != nil {
```

Comment:
Thanks! Yes, I noticed that the `DesktopFile` function would need updating to look for the desktop file matching `X-SnapAppName` (or whatever it was called). That was part of the motivation for creating a whole `snapInfo`/`appInfo` rather than just looking in `/var/lib/snapd/desktop/applications/{snap}_{app}.desktop`, plus it's obviously more robust. I believe when `DesktopFile()` is improved to support this, the usage here should Just Work™.